File: install/static/sudocmd.js
/*jsl:import ipa.js */

/*  Authors:
 *    Endi Sukma Dewata <edewata@redhat.com>
 *
 * Copyright (C) 2010 Red Hat
 * see file 'COPYING' for use and warranty information
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* REQUIRES: ipa.js, details.js, search.js, add.js, entity.js */

IPA.sudocmd = function () {

    var that = IPA.entity({
        'name': 'sudocmd'
    });

    that.init = function() {

        var dialog = IPA.sudocmd_add_dialog({
            'name': 'add',
            'title': 'Add New SUDO Command'
        });
        that.add_dialog(dialog);
        dialog.init();

        var facet = IPA.sudocmd_search_facet({
            'name': 'search',
            'label': 'Search'
        });
        that.add_facet(facet);

        facet = IPA.sudocmd_details_facet({
            'name': 'details',
        });
        that.add_facet(facet);

        that.entity_init();
    };

    return that;
};


IPA.add_entity(IPA.sudocmd());


IPA.sudocmd_add_dialog = function (spec) {

    spec = spec || {};

    var that = IPA.add_dialog(spec);

    that.init = function() {

        that.add_field(IPA.text_widget({name:'sudocmd', undo: false}));
        that.add_field(IPA.text_widget({name:'description', undo: false}));

        that.add_dialog_init();
    };

    return that;
};


IPA.sudocmd_search_facet = function (spec) {

    spec = spec || {};

    var that = IPA.search_facet(spec);

    that.init = function() {

        that.create_column({name:'sudocmd', primary_key: true});
        that.create_column({name:'description'});

        that.search_facet_init();
    };

    return that;
};


IPA.sudocmd_details_facet = function (spec) {

    spec = spec || {};

    var that = IPA.details_facet(spec);

    that.init = function() {

        var section = IPA.details_list_section({
            'name': 'general',
            'label': 'General'
        });
        that.add_section(section);

        section.create_field({'name': 'sudocmd'});
        section.create_field({'name': 'description'});

        section = IPA.details_section({
            'name': 'groups',
            'label': 'Groups'
        });
        that.add_section(section);

        var field = IPA.sudocmd_member_sudocmdgroup_table_widget({
            'name': 'memberof',
            'label': 'Groups',
            'other_entity': 'sudocmdgroup',
            'save_values': false
        });
        section.add_field(field);

        that.details_facet_init();
    };

    return that;
};


IPA.sudocmd_member_sudocmdgroup_table_widget = function (spec) {

    spec = spec || {};

    var that = IPA.association_table_widget(spec);

    that.init = function() {

        var column = that.create_column({
            name: 'cn',
            primary_key: true,
            width: '150px'
        });

        column.setup = function(container, record) {
            container.empty();

            var value = record[column.name];
            value = value ? value.toString() : '';

            $('<a/>', {
                'href': '#'+value,
                'click': function (value) {
                    return function() {
                        var state = IPA.tab_state(that.other_entity);
                        state[that.other_entity + '-facet'] = 'details';
                        state[that.other_entity + '-pkey'] = value;
                        $.bbq.pushState(state);
                        return false;
                    };
                }(value),
                'html': value
            }).appendTo(container);
        };

        that.create_column({
            name: 'description',
            label: 'Description',
            width: '150px'
        });

        that.create_adder_column({
            name: 'cn',
            primary_key: true,
            width: '100px'
        });

        that.create_adder_column({
            name: 'description',
            width: '100px'
        });

        that.association_table_widget_init();
    };

    that.get_records = function(on_success, on_error) {

        if (!that.values.length) return;

        var batch = IPA.batch_command({
            'name': that.entity_name+'_'+that.name+'_show',
            'on_success': on_success,
            'on_error': on_error
        });

        for (var i=0; i<that.values.length; i++) {
            var dn = that.values[i];
            var j = dn.indexOf('=');
            var k = dn.indexOf(',');
            var value = dn.substring(j+1, k);

            var command = IPA.command({
                'method': that.other_entity+'_show',
                'args': [value],
                'options': {
                    'all': true,
                    'rights': true
                }
            });

            batch.add_command(command);
        }

        batch.execute();
    };

    that.add = function(values, on_success, on_error) {

        if (!values.length) return;

        var batch = IPA.batch_command({
            'name': that.entity_name+'_'+that.name+'_add',
            'on_success': on_success,
            'on_error': on_error
        });

        var pkey = $.bbq.getState(that.entity_name + '-pkey', true) || '';

        for (var i=0; i<values.length; i++) {
            var value = values[i];

            var command = IPA.command({
                'method': that.other_entity+'_add_member',
                'args': [value]
            });

            command.set_option('sudocmd', pkey);

            batch.add_command(command);
        }

        batch.execute();
    };

    that.remove = function(values, on_success, on_error) {

        if (!values.length) return;

        var batch = IPA.batch_command({
            'name': that.entity_name+'_'+that.name+'_remove',
            'on_success': on_success,
            'on_error': on_error
        });

        var pkey = $.bbq.getState(that.entity_name + '-pkey', true) || '';

        for (var i=0; i<values.length; i++) {
            var value = values[i];

            var command = IPA.command({
                'method': that.other_entity+'_remove_member',
                'args': [value]
            });

            command.set_option('sudocmd', pkey);

            batch.add_command(command);
        }

        batch.execute();
    };

    return that;
};

File: ipalib/plugins/sudocmd.py
# Authors:
#   Jr Aquino <jr.aquino@citrixonline.com>
#
# Copyright (C) 2010  Red Hat
# see file 'COPYING' for use and warranty information
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import platform
import os
import sys

from ipalib import api, errors, util
from ipalib import Str
from ipalib.plugins.baseldap import *
from ipalib import _, ngettext

__doc__ = _("""
Sudo Commands

Commands used as building blocks for sudo

EXAMPLES:

 Create a new command
   ipa sudocmd-add --desc='For reading log files' /usr/bin/less

 Remove a command
   ipa sudocmd-del /usr/bin/less

""")

topic = ('sudo', _('commands for controlling sudo configuration'))

class sudocmd(LDAPObject):
    """
    Sudo Command object.
    """
    container_dn = api.env.container_sudocmd
    object_name = _('sudo command')
    object_name_plural = _('sudo commands')
    object_class = ['ipaobject', 'ipasudocmd']
    # object_class_config = 'ipahostobjectclasses'
    search_attributes = [
        'sudocmd', 'description',
    ]
    default_attributes = [
        'sudocmd', 'description', 'memberof',
    ]
    attribute_members = {
        'memberof': ['sudocmdgroup'],
    }
    uuid_attribute = 'ipauniqueid'
    label = _('Sudo Commands')
    label_singular = _('Sudo Command')

    takes_params = (
        Str('sudocmd',
            cli_name='command',
            label=_('Sudo Command'),
            primary_key=True,
        ),
        Str('description?',
            cli_name='desc',
            label=_('Description'),
            doc=_('A description of this command'),
        ),
    )

    def get_dn(self, *keys, **options):
        if keys[-1].endswith('.'):
            keys[-1] = keys[-1][:-1]
        dn = super(sudocmd, self).get_dn(*keys, **options)
        try:
            self.backend.get_entry(dn, [''])
        except errors.NotFound:
            try:
                (dn, entry_attrs) = self.backend.find_entry_by_attr(
                    'sudocmd', keys[-1], self.object_class, [''],
                    self.container_dn
                )
            except errors.NotFound:
                pass
        return dn

api.register(sudocmd)

class sudocmd_add(LDAPCreate):
    __doc__ = _('Create new Sudo Command.')

    msg_summary = _('Added Sudo Command "%(value)s"')

api.register(sudocmd_add)

class sudocmd_del(LDAPDelete):
    __doc__ = _('Delete Sudo Command.')

    msg_summary = _('Deleted Sudo Command "%(value)s"')

api.register(sudocmd_del)

class sudocmd_mod(LDAPUpdate):
    __doc__ = _('Modify Sudo Command.')

    msg_summary = _('Modified Sudo Command "%(value)s"')

api.register(sudocmd_mod)

class sudocmd_find(LDAPSearch):
    __doc__ = _('Search for Sudo Commands.')

    msg_summary = ngettext(
        '%(count)d Sudo Command matched', '%(count)d Sudo Commands matched', 0
    )

api.register(sudocmd_find)

class sudocmd_show(LDAPRetrieve):
    __doc__ = _('Display Sudo Command.')

api.register(sudocmd_show)
